interfaces/builtin/udisks2.go

```diff
@@ -0,0 +1,423 @@
+    peer=(label=unconfined),
+
+# Needed for mount/unmount operations
+capability sys_admin,
+
+# Allow reading existing mount points
+@{PROC}/@{pid}/mountinfo r,
+@{PROC}/swaps r,
+
+# Allow scanning of devices
+network netlink raw,
+/run/udev/data/b[0-9]*:[0-9]* r,
+/sys/devices/**/block/** r,
+
+# Mount points could be in /run/media/<user>/* or /media/<user>/*
+/run/systemd/seats/* r,
+/{,run/}media/{,**} rw,
+mount /dev/{sd*,mmcblk*} -> /{,run/}media/**,
+umount /{,run/}media/**,
+
+# These should probably be patched to use $SNAP_DATA/run/...
+/run/udisks2/ rw,
+/run/udisks2/** rw,
+
+# udisksd execs mount/umount to do the actual operations
+/bin/mount ixr,
+/bin/umount ixr,
+
+# mount/umount (via libmount) track some mount info in these files
+/run/mount/utab wrl,
+/run/mount/utab.* wrl,
+
+/dev/sd* r,
+/dev/mmcblk* r,
+`
+
+var udisks2ConnectedSlotAppArmor = []byte(`
+# Allow connected clients to interact with the service. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+dbus (send)
+    bus=system
+    path=/org/freedesktop/UDisks2
+    interface=org.freedesktop.DBus.Properties
+    member=PropertiesChanged
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/UDisks2
+    interface=org.freedesktop.DBus.ObjectManager
+    peer=(label=###PLUG_SECURITY_TAGS###),
+
+# Allow access to the Udisks2 API
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/UDisks2/**
+    interface=org.freedesktop.UDisks2.*
+    peer=(label=###PLUG_SECURITY_TAGS###),
+`)
+
+var udisks2ConnectedPlugAppArmor = []byte(`
+# Description: Allow using udisks service. Reserved because this gives
+#  privileged access to the service.
+# Usage: reserved
+
+#include <abstractions/dbus-strict>
+
+dbus (receive)
+    bus=system
+    path=/org/freedesktop/UDisks2
+    interface=org.freedesktop.DBus.Properties
+    member=PropertiesChanged
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/UDisks2
+    interface=org.freedesktop.DBus.ObjectManager
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# Allow access to the Udisks2 API
+dbus (receive, send)
+    bus=system
+    path=/org/freedesktop/UDisks2/**
+    interface=org.freedesktop.UDisks2.*
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`)
+
+const udisks2PermanentSlotSecComp = `
+bind
+chown32
+fchown
+fchown32
+fchownat
+lchown
+lchown32
+getsockname
+setsockopt
+mount
+recv
+recvfrom
+recvmsg
+send
+sendmsg
+sendto
+shmctl
+umount
+umount2
+`
+
+const udisks2ConnectedPlugSecComp = `
+getsockname
+recv
+recvfrom
+recvmsg
+send
+sendmsg
+sendto
+`
+
+const udisks2PermanentSlotDBus = `
+<policy user="root">
+    <allow own="org.freedesktop.UDisks2"/>
+</policy>
+`
+
+const udisks2ConnectedPlugDBus = `
+<policy context="default">
+    <deny own="org.freedesktop.UDisks2"/>
+    <allow send_destination="org.freedesktop.UDisks2"/>
+</policy>
```

Comment:
Thinking about this more, this needs to be fine-tuned due to this apparmor rule:
```
# Allow unconfined to talk to us
dbus (receive, send)+    bus=system
    path=/org/freedesktop/UDisks2{,/**}
    interface=org.freedesktop.DBus*
    peer=(label=unconfined),
```

As it stands, when udisks2 is installed, any non-snap process of any UID can talk to udisks which would provide easy privilege escalation. On a classic system we might use policykit to control this access, but all snaps does not have policy kit. If there are APIs that are meant to 

I think this should be changed to something like:
```
const udisks2PermanentSlotDBus = `
<policy user="root">
    <allow own="org.freedesktop.UDisks2"/>
    <allow send_destination="org.freedesktop.UDisks2"/>
</policy>
`

const udisks2ConnectedPlugDBus = `
<policy context="default">
    <deny own="org.freedesktop.UDisks2"/>
    <deny send_destination="org.freedesktop.UDisks2"/>
</policy>
`
```

This is in contrast to location-service/observe, which is meant to be accessed by non-root users.

Note that simply removing the apparmor rule does not actually prevent the privilege escalation since users can change_profile (eg, `aa-exec -p snap.udisk2.udisks -- <something nasty>`).